src/libusermode/printers/printers.cpp

```diff
@@ -133,36 +133,133 @@ std::string StringPrinterInterface::print(drakvuf_t drakvuf, drakvuf_trap_info*
 }
 
-std::string WideStringPrinter::getBuffer(vmi_instance_t vmi, const access_context_t* ctx)
+std::string WideStringPrinter::getBuffer(vmi_instance_t vmi, const access_context_t* ctx) const
 {
     auto str_obj = drakvuf_read_wchar_string(vmi, ctx);
     return str_obj == NULL ? "" : (char*)str_obj->contents;
 }
 
-std::string UnicodePrinter::getBuffer(vmi_instance_t vmi, const access_context_t* ctx)
+std::string UnicodePrinter::print(drakvuf_t drakvuf, drakvuf_trap_info* info, uint64_t argument) const
 {
-    auto str_obj = drakvuf_read_unicode_common(vmi, ctx);
-    return str_obj == NULL ? "" : (char*)str_obj->contents;
+    auto vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = argument
+    };
+    std::string str;
+    if (drakvuf_is_wow64(drakvuf, info))
+    {
+        struct
+        {
+            uint16_t length;
+            uint16_t max_length;
+            uint32_t buffer;
+        } __attribute__((packed)) us;
+
+        if (VMI_SUCCESS == vmi_read(vmi, &ctx, sizeof(us), &us, nullptr))
+        {
+            ctx.addr = us.buffer;
+            auto str_obj = drakvuf_read_wchar_string(vmi, &ctx);
+            str = str_obj == NULL ? "" : (char*)str_obj->contents;
+        }
+    }
+    else
+    {
+        auto str_obj = drakvuf_read_unicode_common(vmi, &ctx);
+        str = str_obj == NULL ? "" : (char*)str_obj->contents;
+    }
+    drakvuf_release_vmi(drakvuf);
+    std::stringstream stream;
+    stream << "0x" << std::hex << argument << ":\"" << str << "\"";
+    return stream.str();
+}
+
+std::string PointerToPointerPrinter::print(drakvuf_t drakvuf, drakvuf_trap_info* info, uint64_t argument) const
+{
+    auto vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = argument
+    };
+    addr_t value = 0;
+    if (drakvuf_is_wow64(drakvuf, info))
+        vmi_read_32(vmi, &ctx, reinterpret_cast<uint32_t*>(&value));
+    else
+        vmi_read_addr(vmi, &ctx, &value);
+    drakvuf_release_vmi(drakvuf);
+    std::stringstream stream;
+    stream << "0x" << std::hex << value;
+    return stream.str();
+}
+
+std::string UlongPrinter::print(drakvuf_t drakvuf, drakvuf_trap_info* info, uint64_t argument) const
+{
+    auto vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = argument
+    };
+    uint32_t value = 0;
+    vmi_read_32(vmi, &ctx, &value);
+    drakvuf_release_vmi(drakvuf);
+    std::stringstream stream;
+    stream << "0x" << std::hex << value;
+    return stream.str();
+}
+
+std::string GuidPrinter::print(drakvuf_t drakvuf, drakvuf_trap_info* info, uint64_t argument) const
+{
+    auto vmi = drakvuf_lock_and_get_vmi(drakvuf);
```

Comment:
Use `auto vmi = vmi_lock_guard(drakvuf);` instead.

Reply:
@disaykin `libusermode` is not part of `plugins` where `vmi_lock_guard` defined.

Reply:
We should probably move it out to a `helper.hpp` so that it can be used in `libusermode` as well.

Reply:
@tklengyel can we do this in separate branch to leave this branch more simple?

I have tried to gather here only simple fixes with low diff.

Reply:
Sure